server/src/main/java/timely/Server.java

```diff
@@ -186,74 +187,69 @@ private void shutdownHook() {
     }
 
     public void shutdown() {
+        List<ChannelFuture> channelFutures = new ArrayList<>();
+
+        LOG.info("Closing tcpChannelHandle");
+        channelFutures.add(tcpChannelHandle.close());
+        LOG.info("Closing httpChannelHandle");
+        channelFutures.add(httpChannelHandle.close());
+        LOG.info("Closing wsChannelHandle");
+        channelFutures.add(wsChannelHandle.close());
+        LOG.info("Closing udpChannelHandle");
+        channelFutures.add(udpChannelHandle.close());
+
+        // wait for the channels to shutdown
+        channelFutures.forEach(f -> {
+            try {
+                f.get();
+            } catch (final Exception e) {
+                LOG.error("Error while shutting down channel: {}", f.channel().config());
+                LOG.error("{}", e.getMessage());
+                e.printStackTrace();
+            }
+        });
+
+        List<Future<?>> groupFutures = new ArrayList<>();
+
+        LOG.info("Shutting down tcpBossGroup");
+        groupFutures.add(tcpBossGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down tcpWorkerGroup");
+        groupFutures.add(tcpWorkerGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down httpBossGroup");
+        groupFutures.add(httpBossGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down httpWorkerGroup");
+        groupFutures.add(httpWorkerGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down wsBossGroup");
+        groupFutures.add(wsBossGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down wsWorkerGroup");
+        groupFutures.add(wsWorkerGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down udpBossGroup");
+        groupFutures.add(udpBossGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+        LOG.info("Shutting down udpWorkerGroup");
+        groupFutures.add(udpWorkerGroup.shutdownGracefully(0, 2, TimeUnit.SECONDS));
+
+        groupFutures.forEach(f -> {
```

Comment:
I made this a parallel stream and created a configuration item to set it to 5 seconds by default, and I made it 0 seconds for the test cases. For the default case the parallel stream could make sense because they could take a while to finish and it didn't seem to slow down the 0 wait case.